File: java/client/java/convex/LiveClient.java
package convex;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.WebSocket;
import java.time.Duration;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.BlockingDeque;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.Executors;
import java.util.concurrent.LinkedBlockingDeque;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

/** Native implementation of the pinned convex-rs 0.10.4 unversioned sync profile. */
public final class LiveClient implements AutoCloseable, WebSocket.Listener {
  private static final String INITIAL_TIMESTAMP = "AAAAAAAAAAA=";
  private static final long INITIAL_BACKOFF_MILLIS = 100;
  private static final long MAX_BACKOFF_MILLIS = 15_000;

  private final URI endpoint;
  private final HttpClient http = HttpClient.newHttpClient();
  private final ScheduledExecutorService scheduler =
      Executors.newSingleThreadScheduledExecutor(
          r -> {
            Thread thread = new Thread(r, "convex-java-live");
            thread.setDaemon(true);
            return thread;
          });
  private final Map<Integer, Subscription> subscriptions = new LinkedHashMap<>();
  private int nextId;
  private int querySetVersion;
  private int connectionCount;
  private long reconnectBackoffMillis = INITIAL_BACKOFF_MILLIS;
  private String lastCloseReason = "InitialConnect";
  private String maxObservedTimestamp;
  private JsonNode remoteVersion = zeroVersion();
  private WebSocket socket;
  private boolean reconnectScheduled;
  private boolean closed;
  private final StringBuilder frames = new StringBuilder();

  public LiveClient(String deploymentUrl) {
    URI base = URI.create(deploymentUrl.replaceAll("/+$", ""));
    String scheme = "https".equals(base.getScheme()) ? "wss" : "ws";
    endpoint = URI.create(scheme + "://" + base.getAuthority() + base.getPath() + "/api/sync");
  }

  public synchronized Subscription subscribe(String path, JsonNode args) throws Exception {
    ensureOpen();
    if (path == null || path.isBlank())
      throw new IllegalArgumentException("Convex function path is required");
    if (args == null || !args.isObject())
      throw new IllegalArgumentException("Convex arguments must be a named JSON object");
    Subscription subscription = new Subscription(this, nextId++, path, args.deepCopy());
    subscriptions.put(subscription.queryId, subscription);
    try {
      if (socket == null) connect();
      else modify(List.of(addModification(subscription)));
      return subscription;
    } catch (Exception error) {
      subscriptions.remove(subscription.queryId);
      throw error;
    }
  }

  private synchronized void unsubscribe(Subscription subscription) throws Exception {
    if (subscriptions.remove(subscription.queryId) == null) return;
    subscription.finish();
    if (socket != null)
      modify(
          List.of(
              ConvexClient.JSON
                  .createObjectNode()
                  .put("type", "Remove")
                  .put("queryId", subscription.queryId)));
  }

  private void connect() throws Exception {
    ensureOpen();
    WebSocket connected =
        http.newWebSocketBuilder()
            .connectTimeout(Duration.ofSeconds(10))
            .header("Convex-Client", "java-0.1.0")
            .buildAsync(endpoint, this)
            .get(10, TimeUnit.SECONDS);
    socket = connected;
    querySetVersion = 0;
    remoteVersion = zeroVersion();

    ObjectNode connect =
        ConvexClient.JSON
            .createObjectNode()
            .put("type", "Connect")
            .put("sessionId", UUID.randomUUID().toString())
            .put("connectionCount", connectionCount)
            .put("lastCloseReason", lastCloseReason)
            .put("clientTs", 0);
    if (maxObservedTimestamp != null) connect.put("maxObservedTimestamp", maxObservedTimestamp);
    send(connect);
    if (!subscriptions.isEmpty()) {
      List<ObjectNode> additions = new ArrayList<>();
      for (Subscription subscription : subscriptions.values())
        additions.add(addModification(subscription));
      modify(additions);
    }
    reconnectBackoffMillis = INITIAL_BACKOFF_MILLIS;
    reconnectScheduled = false;
  }

  private ObjectNode addModification(Subscription subscription) {
    ObjectNode add =
        ConvexClient.JSON
            .createObjectNode()
            .put("type", "Add")
            .put("queryId", subscription.queryId)
            .put("udfPath", subscription.path);
    add.putArray("args").add(subscription.args);
    return add;
  }

  private void modify(List<ObjectNode> modifications) throws Exception {
    ObjectNode message =
        ConvexClient.JSON
            .createObjectNode()
            .put("type", "ModifyQuerySet")
            .put("baseVersion", querySetVersion)
            .put("newVersion", querySetVersion + 1);
    ArrayNode array = message.putArray("modifications");
    modifications.forEach(array::add);
    send(message);
    querySetVersion++;
  }

  private void send(JsonNode message) throws Exception {
    WebSocket active = socket;
    if (active == null) throw new IllegalStateException("Live WebSocket is not connected");
    active.sendText(ConvexClient.JSON.writeValueAsString(message), true).get(10, TimeUnit.SECONDS);
  }

  @Override
  public void onOpen(WebSocket webSocket) {
    webSocket.request(1);
  }

  @Override
  public CompletionStage<?> onText(WebSocket webSocket, CharSequence data, boolean last) {
    synchronized (this) {
      if (webSocket != socket) return CompletableFuture.completedFuture(null);
      frames.append(data);
      if (last) {
        try {
          handle(ConvexClient.JSON.readTree(frames.toString()));
        } catch (Exception error) {
          for (Subscription subscription : subscriptions.values())
            subscription.offer(new Update(null, error, List.of()));
          disconnect("ProtocolError", true);
        } finally {
          frames.setLength(0);
        }
      }
    }
    webSocket.request(1);
    return CompletableFuture.completedFuture(null);
  }

  @Override
  public CompletionStage<?> onClose(WebSocket webSocket, int statusCode, String reason) {
    synchronized (this) {
      if (webSocket == socket) disconnect("ServerClosed:" + statusCode, true);
    }
    return CompletableFuture.completedFuture(null);
  }

  @Override
  public void onError(WebSocket webSocket, Throwable error) {
    synchronized (this) {
      if (webSocket == socket) disconnect("TransportError", true);
    }
  }

  private void handle(JsonNode message) {
    String type = message.path("type").asText();
    if ("Ping".equals(type) || "MutationResponse".equals(type) || "ActionResponse".equals(type))
      return;
    if ("TransitionChunk".equals(type))
      throw new IllegalStateException("TransitionChunk is not supported by the Java demonstration");
    if ("FatalError".equals(type) || "AuthError".equals(type))
      throw new IllegalStateException(type + ": " + message.path("error").asText());
    if (!"Transition".equals(type))
      throw new IllegalStateException("unknown Live message: " + type);
    if (!message.path("startVersion").equals(remoteVersion))
      throw new IllegalStateException("Live transition version mismatch");

    Map<Integer, Update> changed = new LinkedHashMap<>();
    for (JsonNode modification : message.path("modifications")) {
      int queryId = modification.path("queryId").asInt();
      List<String> logs = strings(modification.path("logLines"));
      switch (modification.path("type").asText()) {
        case "QueryUpdated" ->
            changed.put(queryId, new Update(modification.get("value"), null, logs));
        case "QueryFailed" ->
            changed.put(
                queryId,
                new Update(
                    null,
                    new ConvexClient.FunctionException(
                        "query",
                        modification.path("errorMessage").asText(),
                        modification.get("errorData"),
                        logs),
                    logs));
        case "QueryRemoved" -> {}
        default ->
            throw new IllegalStateException(
                "unknown Transition modification: " + modification.path("type").asText());
      }
    }
    remoteVersion = message.path("endVersion").deepCopy();
    maxObservedTimestamp = remoteVersion.path("ts").asText(null);
    for (Map.Entry<Integer, Update> entry : changed.entrySet()) {
      Subscription subscription = subscriptions.get(entry.getKey());
      if (subscription != null) subscription.offer(entry.getValue());
    }
  }

  private static List<String> strings(JsonNode node) {
    if (!node.isArray()) return List.of();
    List<String> values = new ArrayList<>();
    node.forEach(item -> values.add(item.asText()));
    return List.copyOf(values);
  }

  public synchronized void debugDisconnect() {
    ensureOpen();
    if (socket == null) throw new IllegalStateException("Live WebSocket is not connected");
    WebSocket previous = socket;
    socket = null;
    previous.abort();
    connectionCount++;
    lastCloseReason = "DebugDisconnect";
    resetRemoteState();
    // Give an in-flight external mutation a chance to commit before the restored
    // query publishes its first value. This is the normal initial transport
    // backoff, not a special retry loop that could hide a broken connection.
    scheduleReconnect(reconnectBackoffMillis);
  }

  private void disconnect(String reason, boolean reconnect) {
    WebSocket previous = socket;
    socket = null;
    if (previous != null) {
      previous.abort();
      connectionCount++;
    }
    lastCloseReason = reason;
    resetRemoteState();
    if (reconnect) scheduleReconnect(reconnectBackoffMillis);
  }

  private void resetRemoteState() {
    querySetVersion = 0;
    remoteVersion = zeroVersion();
  }

  private void scheduleReconnect(long delayMillis) {
    if (closed || subscriptions.isEmpty() || reconnectScheduled) return;
    reconnectScheduled = true;
    scheduler.schedule(
        () -> {
          synchronized (LiveClient.this) {
            reconnectScheduled = false;
            if (closed || subscriptions.isEmpty() || socket != null) return;
            try {
              connect();
            } catch (Exception error) {
              lastCloseReason = error.getMessage();
              long delay = reconnectBackoffMillis;
              reconnectBackoffMillis = Math.min(MAX_BACKOFF_MILLIS, reconnectBackoffMillis * 2);
              scheduleReconnect(delay);
            }
          }
        },
        delayMillis,
        TimeUnit.MILLISECONDS);
  }

  private static JsonNode zeroVersion() {
    ObjectNode version =
        ConvexClient.JSON
            .createObjectNode()
            .put("querySet", 0)
            .put("identity", 0)
            .put("ts", INITIAL_TIMESTAMP);
    return version;
  }

  private void ensureOpen() {
    if (closed) throw new IllegalStateException("Convex Live client is closed");
  }

  @Override
  public synchronized void close() {
    if (closed) return;
    closed = true;
    WebSocket previous = socket;
    socket = null;
    if (previous != null) previous.sendClose(WebSocket.NORMAL_CLOSURE, "closed");
    subscriptions.values().forEach(Subscription::finish);
    subscriptions.clear();
    scheduler.shutdownNow();
  }

  public record Update(JsonNode value, Exception error, List<String> logs) {}

  public static final class Subscription implements AutoCloseable {
    private static final Update CLOSED =
        new Update(null, new IllegalStateException("Live subscription is closed"), List.of());
    private final LiveClient manager;
    private final int queryId;
    private final String path;
    private final JsonNode args;
    private final BlockingDeque<Update> updates = new LinkedBlockingDeque<>(16);
    private boolean closed;

    Subscription(LiveClient manager, int queryId, String path, JsonNode args) {
      this.manager = manager;
      this.queryId = queryId;
      this.path = path;
      this.args = args;
    }

    void offer(Update update) {
      if (closed) return;
      if (!updates.offerLast(update)) {
        updates.pollFirst();
        updates.offerLast(update);
      }
    }

    public Update nextUpdate(Duration timeout) throws Exception {
      Update update = updates.poll(timeout.toMillis(), TimeUnit.MILLISECONDS);
      if (update == null) throw new TimeoutException("timed out waiting for Live update");
      if (update == CLOSED) throw new IllegalStateException("Live subscription is closed");
      return update;
    }

    public JsonNode next(Duration timeout) throws Exception {
      Update update = nextUpdate(timeout);
      if (update.error != null) throw update.error;
      return update.value;
    }

    private void finish() {
      closed = true;
      updates.clear();
      updates.offer(CLOSED);
    }

    @Override
    public void close() {
      if (closed) return;
      try {
        manager.unsubscribe(this);
      } catch (Exception error) {
        throw new IllegalStateException("failed to unsubscribe Live query", error);
      }
    }
  }
}

File: scala/client/scala/convex/LiveClient.scala
package convex

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.ObjectNode
import java.net.URI
import java.net.http.{HttpClient, WebSocket}
import java.time.Duration
import java.util.UUID
import java.util.concurrent.ArrayBlockingQueue
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionException
import java.util.concurrent.CompletionStage
import java.util.concurrent.ExecutionException
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference
import scala.collection.mutable
import scala.jdk.CollectionConverters.*

/** Scala-owned client for the pinned, deliberately narrow Convex sync profile. */
final class LiveClient(rawUrl: String) extends AutoCloseable, WebSocket.Listener:
  import ConvexClient.*
  import LiveClient.*

  private val base = URI.create(rawUrl.replaceAll("/+$", ""))
  private val endpoint = URI.create(
    s"${if base.getScheme == "https" then "wss" else "ws"}://${base.getAuthority}${base.getPath}/api/sync"
  )
  private val http = HttpClient.newHttpClient()
  private val owner: ScheduledExecutorService =
    Executors.newSingleThreadScheduledExecutor(runnable =>
      val thread = new Thread(runnable, "convex-scala-live-owner")
      thread.setDaemon(true)
      thread
    )
  private val subscriptions = mutable.LinkedHashMap.empty[Int, Subscription]
  private val initialWaiters = mutable.LinkedHashMap.empty[Int, CompletableFuture[Subscription]]
  private var nextId = 0
  private var querySetVersion = 0
  private var connectionCount = 0
  private var reconnectBackoffMillis = initialBackoffMillis
  private var lastCloseReason = "InitialConnect"
  private var maxObservedTimestamp: Option[String] = None
  private var remoteVersion: JsonNode = zeroVersion
  private var socket: Option[WebSocket] = None
  private var connecting: Option[CompletableFuture[WebSocket]] = None
  private var reconnectScheduled = false
  private var intentionalRecovery = false
  private var suppressNextHydration = false
  @volatile private var closed = false
  private val fragments = new StringBuilder

  def subscribe(path: String, args: JsonNode): Subscription =
    if path == null || path.isBlank || args == null || !args.isObject then
      throw new IllegalArgumentException(
        "Live subscriptions require a function path and object args"
      )
    val answer = new CompletableFuture[Subscription]()
    val abandoned = new AtomicBoolean(false)
    val created = new AtomicReference[Subscription]()
    submit {
      try
        ensureOpen()
        if abandoned.get() then
          answer.completeExceptionally(new TimeoutException("timed out waiting for Live handshake"))
        else
          val subscription = new Subscription(this, nextId, path, args.deepCopy())
          created.set(subscription)
          nextId += 1
          subscriptions(subscription.queryId) = subscription
          socket match
            case Some(_) =>
              modify(List(add(subscription)))
              answer.complete(subscription)
            case None =>
              initialWaiters(subscription.queryId) = answer
              // A scheduled reconnect owns disconnected state. Starting a competing
              // initial handshake here could consume its timer and strand every query.
              if connecting.isEmpty && !reconnectScheduled then
                beginConnect(reconnecting = connectionCount > 0)
      catch
        case error: Exception =>
          Option(created.get()).foreach(discardSubscription(_, answer))
          answer.completeExceptionally(error)
    }
    try awaitMillis(answer, subscribeTimeoutMillis, "timed out waiting for Live handshake")
    catch
      case error: Throwable =>
        abandoned.set(true)
        val cleanup = new CompletableFuture[Unit]()
        val cleanupScheduled = submit {
          try
            Option(created.get()).foreach(discardSubscription(_, answer))
            cleanup.complete(())
          catch case cleanupError: Throwable => cleanup.completeExceptionally(cleanupError)
        }
        if cleanupScheduled then
          try await(cleanup, 3, "timed out cancelling failed Live subscription")
          catch case cleanupError: Throwable => error.addSuppressed(cleanupError)
        throw error

  private[convex] def unsubscribe(subscription: Subscription): Unit =
    onOwner {
      if subscriptions.remove(subscription.queryId).nonEmpty then
        subscription.finish()
        if socket.nonEmpty then
          modify(
            List(
              ConvexClient.json
                .createObjectNode()
                .put("type", "Remove")
                .put("queryId", subscription.queryId)
            )
          )
    }

  /** Test-only adapter hook. It is deliberately absent from the public facade. */
  private[convex] def debugDisconnect(): Unit =
    onOwner {
      ensureOpen()
      val previous = socket.getOrElse(
        throw new IllegalStateException("Live WebSocket is not connected")
      )
      socket = None
      previous.abort()
      connectionCount += 1
      lastCloseReason = "DebugDisconnect"
      intentionalRecovery = true
      suppressNextHydration = true
      resetRemoteState()
      scheduleReconnect(reconnectBackoffMillis)
      ownerEvent("debugDisconnectScheduled")
    }

  private def beginConnect(reconnecting: Boolean): Unit =
    ensureOpen()
    ownerEvent(if reconnecting then "reconnectBegin" else "initialConnectBegin")
    val future = http
      .newWebSocketBuilder()
      .connectTimeout(Duration.ofSeconds(3))
      .header("Convex-Client", "scala-0.1.0")
      .buildAsync(endpoint, this)
    connecting = Some(future)
    future.whenComplete((connected, failure) =>
      submit {
        if connecting.contains(future) then
          connecting = None
          if closed then
            if connected != null then connected.abort()
            failInitialWaiters(new IllegalStateException("Convex Live client is closed"))
          else if failure != null then
            val error = unwrap(failure)
            if reconnecting then
              connectionCount += 1
              lastCloseReason = "HandshakeError"
              scheduleNextReconnect(error)
            else failInitialWaiters(new TransportError("live", error))
          else finishConnect(connected, reconnecting)
      }
    )

  private def finishConnect(connected: WebSocket, reconnecting: Boolean): Unit =
    try
      socket = Some(connected)
      querySetVersion = 0
      remoteVersion = zeroVersion
      val connect = ConvexClient.json
        .createObjectNode()
        .put("type", "Connect")
        .put("sessionId", UUID.randomUUID().toString)
        .put("connectionCount", connectionCount)
        .put("lastCloseReason", lastCloseReason)
        .put("clientTs", 0)
      maxObservedTimestamp.foreach(connect.put("maxObservedTimestamp", _))
      send(connect)
      if subscriptions.nonEmpty then modify(subscriptions.values.map(add).toList)
      reconnectBackoffMillis = initialBackoffMillis
      reconnectScheduled = false
      ownerEvent(if reconnecting then "reconnectHandshake" else "initialHandshake")
      initialWaiters.foreach { case (queryId, waiter) =>
        subscriptions.get(queryId).foreach(waiter.complete)
      }
      initialWaiters.clear()
    catch
      case error: Exception =>
        disconnect(connected, "TransportError", reconnecting = true, Some(error))
        if !reconnecting then failInitialWaiters(error)

  private def add(subscription: Subscription): ObjectNode =
    val message = ConvexClient.json
      .createObjectNode()
      .put("type", "Add")
      .put("queryId", subscription.queryId)
      .put("udfPath", subscription.path)
    message.putArray("args").add(subscription.args)
    message

  private def modify(modifications: List[ObjectNode]): Unit =
    val message = ConvexClient.json
      .createObjectNode()
      .put("type", "ModifyQuerySet")
      .put("baseVersion", querySetVersion)
      .put("newVersion", querySetVersion + 1)
    val array = message.putArray("modifications")
    modifications.foreach(array.add)
    send(message)
    querySetVersion += 1
    ownerEvent("querySetModified")

  private def send(message: JsonNode): Unit =
    val active = socket.getOrElse(
      throw new IllegalStateException("Live WebSocket is not connected")
    )
    try
      awaitMillis(
        sendText(active, ConvexClient.json.writeValueAsString(message)),
        writeTimeoutMillis,
        "timed out writing Live message"
      )
    catch
      case error: Throwable =>
        val transport = new TransportError("live write", error)
        // Once a write future fails or times out, the peer may have consumed any
        // prefix. Retire that socket instead of guessing at its query-set version.
        disconnect(active, "TransportError", reconnecting = true, Some(transport))
        throw transport

  override def onOpen(webSocket: WebSocket): Unit = submit {
    webSocket.request(1)
  }

  override def onText(
      webSocket: WebSocket,
      data: CharSequence,
      last: Boolean
  ): CompletionStage[?] =
    submit {
      if socket.contains(webSocket) then
        fragments.append(data)
        if last then
          try handle(webSocket, ConvexClient.json.readTree(fragments.toString))
          catch
            case error: ProtocolError =>
              disconnect(webSocket, "ProtocolError", reconnecting = true, Some(error))
            case error: Exception =>
              disconnect(
                webSocket,
                "ProtocolError",
                reconnecting = true,
                Some(new ProtocolError(Option(error.getMessage).getOrElse("invalid Live message")))
              )
          finally fragments.clear()
        if socket.contains(webSocket) then webSocket.request(1)
    }
    CompletableFuture.completedFuture(null)

  override def onClose(
      webSocket: WebSocket,
      statusCode: Int,
      reason: String
  ): CompletionStage[?] =
    submit {
      if socket.contains(webSocket) then
        disconnect(
          webSocket,
          s"ServerClosed:$statusCode",
          reconnecting = true,
          Some(
            new TransportError(
              "live",
              new IllegalStateException(s"Live WebSocket closed: $statusCode $reason")
            )
          )
        )
    }
    CompletableFuture.completedFuture(null)

  override def onError(webSocket: WebSocket, error: Throwable): Unit = submit {
    if socket.contains(webSocket) then
      disconnect(
        webSocket,
        "TransportError",
        reconnecting = true,
        Some(new TransportError("live", unwrap(error)))
      )
  }

  private def handle(webSocket: WebSocket, message: JsonNode): Unit =
    message.path("type").asText match
      case "Ping" | "MutationResponse" | "ActionResponse" => ()
      case "TransitionChunk" =>
        throw new ProtocolError("TransitionChunk is not supported by this demonstration")
      case "FatalError" | "AuthError" =>
        throw new ProtocolError(message.path("error").asText("Live server error"))
      case "Transition" => handleTransition(webSocket, message)
      case other        => throw new ProtocolError(s"unknown Live message: $other")

  private def handleTransition(webSocket: WebSocket, message: JsonNode): Unit =
    if !socket.contains(webSocket) then return
    if !message.path("startVersion").equals(remoteVersion) then
      throw new ProtocolError("Live transition version mismatch")
    if !message.path("endVersion").isObject then
      throw new ProtocolError("Live transition omitted endVersion")
    if !message.path("modifications").isArray then
      throw new ProtocolError("Live transition omitted modifications")

    val changed = mutable.ArrayBuffer.empty[(Int, Update)]
    message.path("modifications").elements().asScala.foreach { modification =>
      if !modification.path("queryId").canConvertToInt then
        throw new ProtocolError("Live modification omitted queryId")
      val queryId = modification.path("queryId").asInt()
      val logs = strings(modification.path("logLines"))
      modification.path("type").asText match
        case "QueryUpdated" =>
          if !modification.has("value") then throw new ProtocolError("QueryUpdated omitted value")
          changed += queryId -> Update(modification.get("value"), null, logs)
        case "QueryFailed" =>
          changed += queryId -> Update(
            null,
            new FunctionError(
              "query",
              modification.path("errorMessage").asText("query failed"),
              modification.get("errorData"),
              logs
            ),
            logs
          )
        case "QueryRemoved" => ()
        case other =>
          throw new ProtocolError(s"unknown Transition modification: $other")
    }

    remoteVersion = message.path("endVersion").deepCopy()
    maxObservedTimestamp = Option(remoteVersion.path("ts").asText(null))
    reconnectBackoffMillis = initialBackoffMillis
    var restoredActiveQuery = false
    changed.foreach { case (queryId, update) =>
      subscriptions.get(queryId).foreach { subscription =>
        val unchangedHydration =
          suppressNextHydration && update.error == null && subscription.sameAsLast(update.value)
        if !unchangedHydration then subscription.offer(update)
        restoredActiveQuery = true
      }
    }
    if restoredActiveQuery then
      intentionalRecovery = false
      suppressNextHydration = false
    ownerEvent("transitionCommitted")

  private def strings(node: JsonNode): List[String] =
    if node.isArray then node.elements().asScala.map(_.asText()).toList else Nil

  private def disconnect(
      expected: WebSocket,
      reason: String,
      reconnecting: Boolean,
      failure: Option[Exception]
  ): Unit =
    if !socket.contains(expected) then return
    socket = None
    expected.abort()
    connectionCount += 1
    lastCloseReason = reason
    suppressNextHydration = true
    resetRemoteState()
    if !intentionalRecovery then
      failure.foreach(error => subscriptions.values.foreach(_.offer(Update(null, error, Nil))))
    if reconnecting then scheduleReconnect(reconnectBackoffMillis)
    ownerEvent("disconnected")

  private def resetRemoteState(): Unit =
    querySetVersion = 0
    remoteVersion = zeroVersion
    fragments.clear()

  private def scheduleReconnect(delayMillis: Long): Unit =
    if !closed && subscriptions.nonEmpty && !reconnectScheduled && connecting.isEmpty then
      reconnectScheduled = true
      reconnectAttemptScheduled(delayMillis)
      owner.schedule(
        new Runnable:
          override def run(): Unit =
            reconnectScheduled = false
            if !closed && subscriptions.nonEmpty && socket.isEmpty && connecting.isEmpty then
              beginConnect(reconnecting = true)
        ,
        delayMillis,
        TimeUnit.MILLISECONDS
      )

  private def scheduleNextReconnect(error: Throwable): Unit =
    ownerEvent("reconnectHandshakeFailed")
    reconnectBackoffMillis = math.min(maxBackoffMillis, reconnectBackoffMillis * 2)
    val delay = reconnectBackoffMillis
    lastCloseReason = Option(error.getMessage).getOrElse("HandshakeError")
    scheduleReconnect(delay)

  private def failInitialWaiters(error: Throwable): Unit =
    initialWaiters.foreach { case (queryId, waiter) =>
      subscriptions.remove(queryId).foreach(_.finish())
      waiter.completeExceptionally(error)
    }
    initialWaiters.clear()

  private def discardSubscription(
      subscription: Subscription,
      waiter: CompletableFuture[Subscription]
  ): Unit =
    initialWaiters.get(subscription.queryId).filter(_ eq waiter).foreach { _ =>
      initialWaiters.remove(subscription.queryId)
    }
    if subscriptions.remove(subscription.queryId).nonEmpty then
      subscription.finish()
      if socket.nonEmpty then
        try
          modify(
            List(
              ConvexClient.json
                .createObjectNode()
                .put("type", "Remove")
                .put("queryId", subscription.queryId)
            )
          )
        catch
          // send() already retired the uncertain socket and scheduled recovery.
          case _: Exception => ()

  private def ensureOpen(): Unit =
    if closed then throw new IllegalStateException("Convex Live client is closed")

  private def submit(action: => Unit): Boolean =
    try
      owner.execute(() => action)
      true
    catch case _: java.util.concurrent.RejectedExecutionException => false

  private def onOwner[T](action: => T): T =
    val answer = new CompletableFuture[T]()
    submit {
      try answer.complete(action)
      catch case error: Throwable => answer.completeExceptionally(error)
    }
    await(answer, 3, "timed out waiting for Live owner")

  override def close(): Unit =
    if closed then return
    onOwner {
      closed = true
      reconnectScheduled = false
      connecting.foreach(_.cancel(true))
      connecting = None
      socket.foreach(_.abort())
      socket = None
      failInitialWaiters(new IllegalStateException("Convex Live client is closed"))
      subscriptions.values.foreach(_.finish())
      subscriptions.clear()
      ownerEvent("closed")
    }
    owner.shutdownNow()

object LiveClient:
  final case class Update(value: JsonNode, error: Exception, logs: List[String])

  private val Closed = Update(null, new IllegalStateException("Live subscription is closed"), Nil)
  @volatile private[convex] var testInitialBackoffMillis = 100L
  @volatile private[convex] var testMaxBackoffMillis = 15_000L
  @volatile private[convex] var testSubscribeTimeoutMillis = 5_000L
  @volatile private[convex] var testWriteTimeoutMillis = 3_000L
  @volatile private[convex] var testReconnectScheduled: Long => Unit = _ => ()
  @volatile private[convex] var testOwnerEvent: (String, String) => Unit = (_, _) => ()
  @volatile private[convex] var testSendText: (WebSocket, String) => CompletableFuture[WebSocket] =
    (socket, text) => socket.sendText(text, true)

  private def initialBackoffMillis: Long = testInitialBackoffMillis
  private def maxBackoffMillis: Long = testMaxBackoffMillis
  private def subscribeTimeoutMillis: Long = testSubscribeTimeoutMillis
  private def writeTimeoutMillis: Long = testWriteTimeoutMillis
  private def reconnectAttemptScheduled(delay: Long): Unit = testReconnectScheduled(delay)
  private def sendText(socket: WebSocket, text: String): CompletableFuture[WebSocket] =
    testSendText(socket, text)
  private def ownerEvent(event: String): Unit =
    testOwnerEvent(event, Thread.currentThread().getName)

  private def zeroVersion: JsonNode = ConvexClient.json
    .createObjectNode()
    .put("querySet", 0)
    .put("identity", 0)
    .put("ts", "AAAAAAAAAAA=")

  private def unwrap(error: Throwable): Throwable = error match
    case completion: CompletionException if completion.getCause != null => completion.getCause
    case execution: ExecutionException if execution.getCause != null    => execution.getCause
    case other                                                          => other

  private def await[T](
      future: java.util.concurrent.Future[T],
      seconds: Long,
      timeoutMessage: String
  ): T =
    try future.get(seconds, TimeUnit.SECONDS)
    catch
      case timeout: TimeoutException =>
        throw new TimeoutException(timeoutMessage)
      case execution: ExecutionException => throw unwrap(execution)

  private def awaitMillis[T](
      future: java.util.concurrent.Future[T],
      milliseconds: Long,
      timeoutMessage: String
  ): T =
    try future.get(milliseconds, TimeUnit.MILLISECONDS)
    catch
      case timeout: TimeoutException =>
        throw new TimeoutException(timeoutMessage)
      case execution: ExecutionException => throw unwrap(execution)

  final class Subscription private[convex] (
      manager: LiveClient,
      private[convex] val queryId: Int,
      private[convex] val path: String,
      private[convex] val args: JsonNode
  ) extends AutoCloseable:
    private val updates = new ArrayBlockingQueue[Update](17)
    private var closed = false
    private var lastValue: JsonNode = null

    private[convex] def sameAsLast(value: JsonNode): Boolean = synchronized {
      lastValue != null && lastValue.equals(value)
    }

    private[convex] def offer(update: Update): Unit = synchronized {
      if closed then return
      if update.error == null then lastValue = update.value.deepCopy()
      while updates.size() >= 16 do updates.poll()
      updates.offer(update)
    }

    private[convex] def finish(): Unit = synchronized {
      if closed then return
      closed = true
      updates.offer(Closed)
    }

    def nextUpdate(timeout: Duration): Update =
      val update = updates.poll(timeout.toMillis, TimeUnit.MILLISECONDS)
      if update == null then throw new TimeoutException("timed out waiting for Live update")
      if update eq Closed then throw new IllegalStateException("Live subscription is closed")
      update

    def next(timeout: Duration): JsonNode =
      val update = nextUpdate(timeout)
      if update.error != null then throw update.error
      update.value

    override def close(): Unit =
      val shouldUnsubscribe = synchronized(!closed)
      if shouldUnsubscribe then manager.unsubscribe(this)
